src/ingest/SpecRepo.cpp

```diff
@@ -169,30 +150,70 @@ void genSpecs4Roll(const std::string& version,
     // making a s3 fs with given host
     auto fs = nebula::storage::makeFS("s3", sourceInfo.host);
 
+    // exact macro pattern type
+    auto pt = nebula::meta::extractPattern(table->timeSpec.pattern);
+
     // list all objects/files from given path
     // A roll spec will cover X days given table location of source data
     const auto now = Evidence::now();
     const auto max_days = table->max_hr / 24;
-    size_t i;
-    for (i = 0; i <= max_days; ++i) {
-      // run hourly partition list
-      for(size_t j = 0 ; j < DAY_HOURS; j++) {
-        genSpecs4HourlyRoll(version, table, 24, i, specs);
+    // start spec in ascending order
+    long startTime = now - max_days * DAY_SECONDS - (table->max_hr % 24) * HOUR_SECONDS;
+
+    /**
+     * max_hr could cross day boundary so we starts with a day before max_hr cut-off date
+     * and skip those earlier than startTime, we only generate spec watermark
+     * - in timestamp ascending order
+     * - at leaf dirs match pattern defines as specific granularity (date,hour,minute,second)
+     */
+    for (long day_ago = max_days + 1; day_ago >= 0; day_ago--) {
```

Comment:
I suggest we split each pattern type process into a subroutine instead of in a multi layers of `for blocks`.
Clearly to declare that we only supports 4 different patterns for now, even though we may evolve into more generic framework.

In addition to that, we may consider supporting categorical pattern as well for partitioning by enumerable values,  such as `s3://zzz/app=[a, b, c]/dt={date}/hour={hr}`, in this case, app will have 3 different values `a` or `b` or `c`

Reply:
Updated and remove for loops with iterative approach. Still haven't got chance working on app=[a, b, c] like macro due to the fact we need to support predefined collection macro pattern or on demand list based collection generation requires a bit more work